kernel/CMakeLists.txt

```diff
@@ -39,6 +38,7 @@ target_sources_ifdef(CONFIG_STACK_CANARIES        kernel PRIVATE compiler_stack_
 target_sources_ifdef(CONFIG_SYS_CLOCK_EXISTS      kernel PRIVATE timeout.c timer.c)
 target_sources_ifdef(CONFIG_ATOMIC_OPERATIONS_C   kernel PRIVATE atomic_c.c)
 target_sources_if_kconfig(                        kernel PRIVATE poll.c)
+target_sources_if_kconfig(                        kernel PRIVATE mempool.c)
```

Comment:
Instead of selectively including the C file, another option would be to put an #ifdef around the SYS_INIT() macro in mempool.c, which is the actual symbol that is sucking in the code you're worried about.